### DISCUSSION TITLE
tanka tutorial working with jsonnet-libs

### DISCUSSION THREAD

**Discussion by ludzzz** [grafana/tanka#481] [Q&A]
Hello,

I am trying to use tanka to organize/manage the Kubernetes setup.
I tried to follow: https://tanka.dev/ and https://grafana.com/docs/loki/latest/installation/tanka/
My goal is to have a setup for Grafana, Prometheus, and Loki that I can use to deploy on different clusters.

But I am a bit confused with all the resources and different libraries.

1. First ksonnet:

What is the difference between https://github.com/ksonnet/ksonnet-lib and https://github.com/jsonnet-libs/k8s-alpha
My understanding is the ksonnet-lib is discontinued and only cover until Kubernetes 1.14, and jsonnet-libs/k8s-alpha is the attempt to replace it.

2. How to do the tutorial of tanka.dev with github.com/jsonnet-libs/k8s-alpha (1.16 in my case)?

In my case I am importing:
```
(import 'github.com/jsonnet-libs/k8s-alpha/1.16/main.libsonnet') +
(import 'github.com/jsonnet-libs/k8s-alpha/1.16/extensions/kausal-shim.libsonnet')
```
and for grafana doing:
```
{
  local deploy = $.apps.v1.deployment,
  local container = $.core.v1.container,
  local port = $.core.v1.containerPort,
  local service = $.core.v1.service,

  grafana: {
    deployment: deploy.new(name=$._config.grafana.name, replicas=1, containers=[
      // container constructor
      container.new($._config.grafana.name, 'grafana/grafana')
      + container.withPorts(
        [port.new('ui', $._config.grafana.port)]
      ),
    ]),
    service: $.util.serviceFor(self.deployment)
             + service.mixin.spec.withType('NodePort'),
  },
}
```
getting the error: 
```
evaluating jsonnet: RUNTIME ERROR: function expected 1 positional argument(s), but got 2
	/home/myself/tanka/monitoring2/environments/default/grafana.jsonnet:12:10-48	thunk from <thunk from <thunk from <thunk from <object <anonymous>>>>>
	During manifestation	
```
Any help would be much appreciated.
Thank you

**Comment by ludzzz** [comment]
Starting to get my head around it.
To fix the container port I did:
```
{
  local deploy = $.apps.v1.deployment,
  local container = $.core.v1.container,
  local port = $.core.v1.containerPort,
  local service = $.core.v1.service,

  grafana: {
    deployment: deploy.new(name=$._config.grafana.name, replicas=1, containers=[
      // container constructor
      container.new($._config.grafana.name, 'grafana/grafana')
      + container.withPorts(
        [port.new($._config.grafana.port) + port.withName('ui')]
      ),
    ]),
  },
}
```
Don't forget to import k8s-alpha.

One thing overall, I think the tanka.dev tutorial is deprecated.
It would be a good thing to redo it with k8s-alpha library.
Is there a way to propose a PR for it?

**Comment by malcolmholmes** [comment]
You are right that the tanka.dev tutorial needs updating. In fact  we need to complete the process of migrating from ksonnet-lib to k8s-alpha (which involves renaming this to something without 'alpha' in its name). Some progress was made here just recently. I believe you no longer need the kausal-shim in k8s-alpha as of yesterday.

We're reluctant to update the docs until this migration work has been done. However, a PR against the docs in this repo could really help.
